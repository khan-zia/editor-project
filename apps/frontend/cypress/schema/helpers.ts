type StrictProperties = {
  required: true | string[];
  additionalProperties: boolean;
};

export const strictProperties: StrictProperties = {
  // Mark all properties inside the "schema" block as required.
  // Those all must be sent in the payload by the frontend.
  required: true,

  // Do not expect any extra properties that are not mentioned inside the "schema" block
  // to be present in the payload sent by the frontend.
  additionalProperties: false,
};
