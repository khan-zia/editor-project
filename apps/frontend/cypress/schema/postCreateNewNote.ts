import { ObjectSchema, versionSchemas } from '@cypress/schema-tools';
import { strictProperties } from './helpers';
import { Formats } from '../formats';

type Request = {
  title: string;
};

const RequestExample: Request = {
  title: 'Awesome Note',
};

const Request: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'CreateNewNoteRequest',
    type: 'object',
    description: 'The payload sent to the backend creating a new note.',
    properties: {
      title: {
        type: 'string',
        description: 'Title of the note.',
      },
    },
    ...strictProperties,
  },
  example: RequestExample,
};

// What will the backend return if creating a note fails.
// typically an object but our backend returns just "null" for this demo.
type NoteCreationFailureResponse = null;

const NoteCreationFailureExample: NoteCreationFailureResponse = null;

const NoteCreationFailureResponse: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'CreateNewNoteFailure',
    type: null,
    description: 'Response of the backend when creating a new note fails.',
    ...strictProperties,
  },
  example: NoteCreationFailureExample,
};

// When the backend succeeds creating a new note, what will the response look like?
// It returns ID and title of the newly created note.
type NoteCreationSuccessResponse = {
  id: string;
  title: string;
};

const SuccessResponseExample: NoteCreationSuccessResponse = {
  id: '1JrclVI6sP3zw03xs3m7',
  title: 'Awesome Note',
};

export const NoteCreationSuccessResponse: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'CreateNewNoteSuccess',
    description: 'Payload of the backend response upon successful creation of a new note.',
    type: 'object',
    properties: {
      id: {
        type: 'string',
        description: 'ID of the newly created noted.',
        format: Formats.NoteID.name, // This will check the ID against the regex defined in the format.
      },
      title: {
        type: 'string',
        description: 'Title of the newly created note.',
      },
    },
    ...strictProperties,
  },
  example: SuccessResponseExample,
};

export const CreateNewNoteRequest = versionSchemas(Request);
export const CreateNewNoteFailure = versionSchemas(NoteCreationFailureResponse);
export const CreateNewNoteSuccess = versionSchemas(NoteCreationSuccessResponse);
