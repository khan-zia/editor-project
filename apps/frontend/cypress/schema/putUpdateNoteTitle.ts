import { ObjectSchema, versionSchemas } from '@cypress/schema-tools';
import { strictProperties } from './helpers';

type Request = {
  title: string;
};

const RequestExample: Request = {
  title: 'Awesome Note',
};

const Request: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'UpdateTitleRequest',
    type: 'object',
    description: "The payload sent to the backend for updating a note's title.",
    properties: {
      title: {
        type: 'string',
        description: 'New title of the note.',
      },
    },
    ...strictProperties,
  },
  example: RequestExample,
};

type UpdateTitleFailureResponse = {
  success: false;
};

// This example will be shown in the cypress console in case backend returns
// an expected value.
const UpdateTitleFailureExample: UpdateTitleFailureResponse = {
  success: false,
};

const UpdateTitleFailureResponse: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'UpdateTitleFailure',
    type: null,
    description: 'Response of the backend when updating a note title fails.',
    properties: {
      success: {
        type: 'boolean',
        description: 'Status of the update title request.',
      },
    },
    ...strictProperties,
  },
  example: UpdateTitleFailureExample,
};

// When a title is successfully updated, what will the backend's response look like?
// It simply returns "true" for the purpose of this test.
type UpdateTitleSuccessResponse = {
  success: true;
};

const UpdateTitleSuccessExample: UpdateTitleSuccessResponse = {
  success: true,
};

export const UpdateTitleSuccessResponse: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'UpdateTitleSuccess',
    description: 'Payload of the backend response when a note title is successfully updated.',
    type: 'object',
    properties: {
      success: {
        type: 'boolean',
        description: 'Status of the update title request.',
      },
    },
    ...strictProperties,
  },
  example: UpdateTitleSuccessExample,
};

export const UpdateTitleRequest = versionSchemas(Request);
export const UpdateTitleFailure = versionSchemas(UpdateTitleFailureResponse);
export const UpdateTitleSuccess = versionSchemas(UpdateTitleSuccessResponse);
