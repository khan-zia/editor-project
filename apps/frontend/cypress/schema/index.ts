import { SchemaCollection, bind, combineSchemas } from '@cypress/schema-tools';
import { CreateNewNoteRequest, CreateNewNoteFailure, CreateNewNoteSuccess } from './postCreateNewNote';
import { Formats as formats } from '../formats';
import { UpdateTitleFailure, UpdateTitleRequest, UpdateTitleSuccess } from './putUpdateNoteTitle';

export const schemas: SchemaCollection = combineSchemas(
  CreateNewNoteRequest,
  CreateNewNoteFailure,
  CreateNewNoteSuccess,
  UpdateTitleRequest,
  UpdateTitleFailure,
  UpdateTitleSuccess,
);

export const api = bind({ schemas, formats });
