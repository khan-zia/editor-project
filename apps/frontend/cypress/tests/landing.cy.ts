import { api } from '../schema';

describe('Rendering the landing page and being able to create a new note.', () => {
  it('Can render the landing page properly.', () => {
    // Visit the main page.
    cy.visit('/');

    // The "Notes" word should be there in the top left corner.
    cy.getByDataAttribute('notes-logo').canBeSeen();

    // The button for adding a new note should be there.
    cy.getByDataAttribute('add-note-button').canBeSeen();
  });

  it('Can create a new note.', () => {
    // This is a patch for the snapshot package. Has nothing to do with the test.
    cy.fixCypressSpec();

    // Visit the main page.
    cy.visit('/');

    // Spy on the new note creation request.
    cy.intercept(
      {
        method: 'POST',
        url: 'http://localhost:3001/api/notes',
      },

      /**
       * Make sure correct payload was sent to the backend for creating a new note.
       * This is made possible with JSON schemas. You define a schema for what you expect
       * the payload should be generated by the frontend.
       */
      (request) => api.assertSchema('CreateNewNoteRequest', '1.0.0')(request.body),
    ).as('createNoteRequest');

    // Click the "add note" button.
    cy.getByDataAttribute('add-note-button').click();

    // The new note dialog should be visible.
    cy.getByDataAttribute('new-note-dialog').canBeSeen();

    // There should be a cancel button on the dialog that also closes it.
    cy.getByDataAttribute('close-dialog-button').canBeSeen().click();
    cy.getByDataAttribute('new-note-dialog').should('not.exist');

    // Bring the dialog back.
    cy.getByDataAttribute('add-note-button').click();

    // The input field for the entering a title should be visible.
    cy.getByDataAttribute('new-note-input').canBeSeen();

    // Clicking the create button without entering a title triggers an error message.
    cy.getByDataAttribute('create-note-button').canBeSeen().click();
    cy.getByDataAttribute('note-error-message')
      .canBeSeen()
      .then((element) => {
        expect(element.text()).to.be.eq('Choose a title for your note.');
      });

    // Enter a title for the note.
    cy.getByDataAttribute('new-note-input').type('Awesome Note');

    // Click the create note button again.
    cy.getByDataAttribute('create-note-button').click();

    // Wait for the new note creation request to complete and then run assertions
    // against the response.
    cy.wait('@createNoteRequest')
      .its('response')
      .then(({ body, statusCode }) => {
        expect(statusCode).to.eq(200);

        // Asserts json schema structure. Response body must be the same
        // as the expected json schema.
        api.assertSchema('CreateNewNoteSuccess', '1.0.0')(body);

        // Sanitize response body. Because volatile values can not be deterministically
        // tested. e.g. Note's ID is random and will be different every time. We only
        // validate it against a defined regex format and then we substitute it by an example
        // value. The example value can then be deterministically checked against a stored snapshot.
        const res = api.sanitize('CreateNewNoteSuccess', '1.0.0')(body);

        // Check that the response body matches a store snapshot. If no snapshot exist for this
        // request, it will be created and stored as the first time.
        cy.wrap(res).toMatchSnapshot();
      });

    // The user should land on the new note's page.
    const base = Cypress.config().baseUrl + '/notes/';
    cy.url()
      .should('include', base)
      .then((url) => {
        // New note's ID is expected to be a 20 chars alphanumeric string.
        expect(new RegExp('^[a-zA-Z0-9]{20}$').test(url.replace(base, ''))).to.be.true;
      });
  });
});
