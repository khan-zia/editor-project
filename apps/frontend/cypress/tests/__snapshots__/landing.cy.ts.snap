exports[`Rendering the landing page properly. > Can create a new note. #0`] =
{
  "id": "1JrclVI6sP3zw03xs3m7",
  "title": "Awesome Note"
};

exports[`Rendering the landing page and being able to create a new note. > Can create a new note. #0`] =
{
  "id": "1JrclVI6sP3zw03xs3m7",
  "title": "Awesome Note"
};
